Switch to meson subprojects/wraps

ret2libc opened the pull request:
**DO NOT MERGE YET, see https://github.com/rizinorg/rizin/pull/510**

 **Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

This PR makes use of meson subprojects. It removes git submodules.

Advantages:
- no need to mess with git submodules, --recurse-submodules, etc.
- easy to download external deps, because you don't need to check if tar/wget/curl are available (in Windows they are probably not), so it can work easily on multiple platforms
- reduce the complexity of the meson.build files, because they are more focused on a single subproject
- there are already wraps available in Meson WrapDB for lz4, zlib, libzip, xxhash. Other stuff could be moved to subprojects as well easily.

Possible disadvantages:
- once a subproject is downloaded, it is not automatically updated if you update the .wrap file. You have to either delete the old `subprojects/X` directory or run `meson subproject update`.

ret2libc committed:
Use wget/tar when using capstone v3/v4/v5

ret2libc committed:
Use meson subprojects for external libs

* Provide separate capstone wraps, so you can easily compile against
  different capstone versions

ret2libc committed:
Use rizinorg/capstone when capstone-bundled is selected

We can apply special patches there if necessary.

ret2libc commented:
This closes https://github.com/rizinorg/rizin/issues/188

ret2libc committed:
Fix capstone-next CI and includes of capstone >= 4

XVilka commented:
It looks much better. I will test it locally, and if it works on my setups - let's merge this. 👍

wargio committed:
Merge branch 'dev' into capstone-meson

ret2libc committed:
REMOVE ME: enable appveyor

XVilka commented:
It doesn't solve the problem of "git clone" - it still uses it unconditionally, thus https://github.com/rizinorg/rizin/issues/188 issue still persists.

ret2libc commented:
What command are you running? If you run meson with -Duse_capstone_version=v4 i doubt it uses git clone

XVilka commented:
Yes, this way it works, but there is no way to fetch latest capstone from git repository this way.
What about creating a copy of `subprojects/capstone-bundled.wrap` and `subprojects/capstone-next.wrap` using direct URL like `subprojects/capstone-v4.wrap` does? Call them `subprojects/capstone-bundled-archive.wrap` and `subprojects/capstone-next-archive.wrap` maybe? Same with SDB.

XVilka commented:
The rest is LGTM.

XVilka commented:
But seems there is a problem on MacOS:
```
[44/1485] Linking target subprojects/sdb/libsdb.1.5.1-git.dylib
FAILED: subprojects/sdb/libsdb.1.5.1-git.dylib 
clang  -o subprojects/sdb/libsdb.1.5.1-git.dylib subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_array.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_set.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_base64.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_buffer.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_cdb.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_cdb_make.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_dict.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_diff.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_disk.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_fmt.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_ht_uu.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_ht_pp.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_ht_up.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_journal.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_json.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_lock.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_ls.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_match.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_ns.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_num.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_query.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_sdb.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_sdbht.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_util.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_text.c.o -Wl,-dead_strip_dylibs -Wl,-headerpad_max_install_names -Wl,-undefined,error -shared -install_name @rpath/libsdb.1.5.1-git.dylib -compatibility_version 1.5.1-git -current_version 1.5.1-git
ld: malformed 32-bit x.y.z version number: 1.5.1-git
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

ret2libc commented:
> Yes, this way it works, but there is no way to fetch latest capstone from git repository this way.
> What about creating a copy of `subprojects/capstone-bundled.wrap` and `subprojects/capstone-next.wrap` using direct URL like `subprojects/capstone-v4.wrap` does? Call them `subprojects/capstone-bundled-archive.wrap` and `subprojects/capstone-next-archive.wrap` maybe? Same with SDB.

I understand your problems with the firewall, but IMO we should not go too far and duplicate everything. That means more things to test and we already support a lot of configurations. In your case I would either just use `-Duse_capstone_version=v4` and stick with it or manually download the archive from our repository `rizinorg/capstone` in the `subprojects/` directory.

XVilka commented:
There are millions of people who will need this. Not just me.

ret2libc commented:
> There are millions of people who will need this. Not just me.

Then they should stick with releases probably. We are a project based on `git`, not using `git` at all seems ... weird. If you can't use `git`, you can't develop. If you can't develop, you just stick with releases that just involves downloading a tarball and unzipping it.

XVilka commented:
It's possible to use `git`, it's just cloning `capstone` is slow due to the repository/history size. Anyway, if you don't want to do this - I can send my own PR.

ret2libc committed:
Merge remote-tracking branch 'origin/dev' into capstone-meson

ret2libc committed:
Just use sdb_native_exe

ret2libc commented:
I will need https://github.com/rizinorg/sdb/pull/5 to continue.

ret2libc committed:
Merge branch 'dev' into capstone-meson

ret2libc committed:
Update SDB

XVilka commented:
Seems like broken VS build:
```
OLDNAMES.lib(unlink.obi) : error LNK2001: unresolved external symbol __imp_unlink
librz_analysis.a(src_cdb_make.c.obj) : error LNK2019: unresolved external symbol __imp__aligned_free referenced in function cdb_alloc_free
librz_analysis.a(src_cdb_make.c.obj) : error LNK2019: unresolved external symbol __imp__aligned_malloc referenced in function cdb_alloc
librz_analysis.a(src_query.c.obj) : error LNK2019: unresolved external symbol __imp_puts referenced in function sdb_query
OLDNAMES.lib(unlink.obi) : error LNK2001: unresolved external symbol __imp__unlink
test/unit/test_sign.exe : fatal error LNK1120: 5 unresolved externals
ninja: build stopped: subcommand failed.
Command exited with code 1
```

thestr4ng3r reviewed (approved):
lgtm after rebase and ci fix

ret2libc committed:
Merge remote-tracking branch 'origin/dev' into capstone-meson

ret2libc committed:
Remove capstone_in_builddir meson option

ret2libc committed:
Add script to check if subprojects need to be updated

thestr4ng3r commented on meson.build:
```diff
@@ -415,6 +416,11 @@ if not zlib_dep.found() or not get_option('use_sys_zlib')
 
+r = run_command(py3_exe, check_meson_subproject_py, 'sdb')
+if r.returncode() == 1 and not get_option('no_subprojects_check')
+  error('Subproject sdb is not updated. Please make sure to either run `meson subprojects update`.')
```
either run this or wat

ret2libc replied:
ahh :D old msg !

ret2libc committed:
Remove "either"

ret2libc committed:
Mention no_subprojects_check option in error msg

ret2libc committed:
Merge remote-tracking branch 'origingit/dev' into capstone-meson

ret2libc committed:
CHANGE ME: use pr sdb

ret2libc committed:
Merge branch 'dev' into capstone-meson

ret2libc committed:
Merge remote-tracking branch 'origin/dev' into capstone-meson

ret2libc committed:
Use latest SDB

XVilka commented:
Seems there is a problem:
```
FAILED: librz/cons/rz_cons.dll 
"link"  /MACHINE:x64 /OUT:librz/cons/rz_cons.dll librz/cons/rz_cons.dll.p/canvas.c.obj librz/cons/rz_cons.dll.p/canvas_line.c.obj librz/cons/rz_cons.dll.p/cons.c.obj librz/cons/rz_cons.dll.p/html.c.obj librz/cons/rz_cons.dll.p/editor.c.obj librz/cons/rz_cons.dll.p/grep.c.obj librz/cons/rz_cons.dll.p/hud.c.obj librz/cons/rz_cons.dll.p/input.c.obj librz/cons/rz_cons.dll.p/less.c.obj librz/cons/rz_cons.dll.p/line.c.obj librz/cons/rz_cons.dll.p/more.c.obj librz/cons/rz_cons.dll.p/output.c.obj librz/cons/rz_cons.dll.p/pager.c.obj librz/cons/rz_cons.dll.p/pal.c.obj librz/cons/rz_cons.dll.p/cpipe.c.obj librz/cons/rz_cons.dll.p/rgb.c.obj librz/cons/rz_cons.dll.p/cutf8.c.obj "/nologo" "/release" "/nologo" "/OPT:REF" "/DLL" "/IMPLIB:librz/cons\rz_cons.lib" "librz/util\rz_util.lib" "librzspp.a" "subprojects\sdb\src\libsdb.a" "librzzlib.a" "ws2_32.lib" "wininet.lib" "psapi.lib" "kernel32.lib" "user32.lib" "gdi32.lib" "winspool.lib" "shell32.lib" "ole32.lib" "oleaut32.lib" "uuid.lib" "comdlg32.lib" "advapi32.lib"
libsdb.a(sdb.c.obj) : error LNK2005: sdb_foreach_list already defined in rz_util.lib(rz_util.dll)
libsdb.a(sdb.c.obj) : error LNK2005: sdb_free already defined in rz_util.lib(rz_util.dll)
libsdb.a(sdb.c.obj) : error LNK2005: sdb_new0 already defined in rz_util.lib(rz_util.dll)
libsdb.a(sdb.c.obj) : error LNK2005: sdb_set already defined in rz_util.lib(rz_util.dll)
libsdb.a(ls.c.obj) : error LNK2005: ls_free already defined in rz_util.lib(rz_util.dll)
libsdb.a(ht_pp.c.obj) : error LNK2005: ht_pp_find already defined in rz_util.lib(rz_util.dll)
libsdb.a(ht_pp.c.obj) : error LNK2005: ht_pp_free already defined in rz_util.lib(rz_util.dll)
libsdb.a(ht_pp.c.obj) : error LNK2005: ht_pp_new already defined in rz_util.lib(rz_util.dll)
   Creating library librz/cons\rz_cons.lib and object librz/cons\rz_cons.exp
librz/cons/rz_cons.dll : fatal error LNK1169: one or more multiply defined symbols found
[713/1489] Generating symbol file librz/socket/rz_socket.dll.p/rz_socket.dll.symbols
[714/1489] Generating symbol file librz/crypto/rz_crypto.dll.p/rz_crypto.dll.symbols
ninja: build stopped: subcommand failed.
Command exited with code 1
```
https://ci.appveyor.com/project/rizinorg/rizin/builds/36938607/job/mwcnf204b7b8hknh#L1275

ret2libc committed:
Merge branch 'dev' into capstone-meson

XVilka committed:
Merge branch 'dev' into capstone-meson

XVilka commented:
There are many conflicts now, please rebase when you have time.

ret2libc commented:
> There are many conflicts now, please rebase when you have time.

Yeah I'm waiting for first release, I will revive this after that.

ret2libc committed:
Merge remote-tracking branch 'origin/dev' into capstone-meson

ret2libc committed:
Update sdb revision

ret2libc committed:
update to tree-sitter-0.18.0 and remove -static from meson

ret2libc committed:
Remove git submodules

ret2libc committed:
Remove references to git submodules

ret2libc committed:
Merge remote-tracking branch 'origin/dev' into capstone-meson

ret2libc commented on doc/PACKAGERS.md:
````diff
@@ -6,15 +6,24 @@ Thank you for taking the time to package Rizin!
+$ git clone https://github.com/rizinorg/rizin
+$ cd rizin
+$ meson subprojects download
+```
````
add mention of `meson dist`

ret2libc replied:
Or maybe not, since you have to have anyway a build directory to run `meson dist`.

ret2libc commented on .github/workflows/ci.yml:
```diff
@@ -445,14 +433,12 @@ jobs:
           - name: vs2019_static
             compiler: cl
-            meson_options: --default-library=static -Dstatic_runtime=true
+            meson_options: --default-library=static -Db_vscrt=static_from_buildtype
```
review if it's really necessary `static_runtime`. Maybe just remove in favour of this.

ret2libc commented on meson_options.txt:
```diff
@@ -4,6 +4,7 @@ option('cli', type: 'feature', value: 'auto', description: 'Build CLI programs (
 option('static_runtime', type: 'boolean', value: false, description: 'Set to true when building a static version of Rizin')
 option('local', type: 'feature', value: 'auto', description: 'Adds support for local/side-by-side installation (sets rpath if needed)')
 option('blob', type: 'boolean', value: false, description: 'Compile just one binary which dispatch to the right handlers based on the name used to call it')
+option('no_subprojects_check', type: 'boolean', value: false, description: 'Do not check if git subprojects are up-to-date. Useful when developing on a different subproject version')
```
maybe use affermative option instead of negative one. `option(sub-projects_check, value: true, ..)`

ret2libc committed:
REMOVE ME: tmp move SDB to SDB PR for testing

ret2libc committed:
Use latest SDB

ret2libc committed:
Merge branch 'dev' into capstone-meson

ret2libc committed:
Fix some meson messages

ret2libc committed:
Use subprojects_check instead of no_subprojects_check

ret2libc committed:
Merge branch 'dev' into capstone-meson

08A commented on .github/workflows/clang-analysis.yml:
```diff
@@ -16,8 +16,6 @@ jobs:
     - name: Checkout repository
       uses: actions/checkout@v2
-      with:
-          submodules: recursive
```
You can replace line 54
```
        grep -Rl 'rizin/shlr/capstone' reports | xargs rm -fv
        grep -Rl 'rizin/shlr/sdb' reports | xargs rm -fv
        grep -Rl 'rizin/shlr/tree-sitter' reports | xargs rm -fv
```

with

```
        grep -Rl 'rizin/subprojects' reports | xargs rm -fv
```

ret2libc replied:
Thanks, done!

ret2libc committed:
Remove references to shlr/{capstone,tree-sitter}

kazarmy commented on shlr/meson.build:
```diff
@@ -1,211 +1,35 @@
+      error('Subproject capstone-next is not updated. Please make sure to run `meson subprojects update` or set option `subprojects_check=false` if you want to compile against currently downloaded subproject.')
     endif
-
-    patches_files = [
-      'capstone-calloc.patch',
-      'fix-x86-16.patch',
-      'sparc-crash.patch',
-      'sstream-null.patch'
-    ]
-
-    message('Patching capstone with rizin\'s patches')
-    foreach file : patches_files
-      patch_path = join_paths(meson.current_source_dir(), 'capstone-patches', file)
-      patch_cmd = run_command(py3_exe, git_exe_repo_py, git_exe, capstone_path, 'apply', '-p1', patch_path)
-      if patch_cmd.returncode() != 0
-        warning('Cannot apply patch ' + file)
-      endif
-    endforeach
-  endif
```
We don't need to patch capstone anymore?

ret2libc replied:
See capstone bundled wrap. We have a fork there and i already applied the patches in the fork

kazarmy commented on doc/capstone.md:
```diff
@@ -4,15 +4,16 @@ Capstone
 By default Rizin will build statically against a bundled version of capstone4
-(with some custom patches applied). Please be aware that by default Rizin will
-download Capstone in the source directory `shlr/capstone`.
+(with some custom patches applied). Please be aware that by default Rizin
```
If no patches are applied then the parenthesized phrase can be removed.

ret2libc replied:
i'll keep the parentheses because the patches are still applied, they just are directly in the git repo.

kazarmy reviewed (approved):
lgtm

ret2libc committed:
Merge branch 'dev' into capstone-meson

ret2libc committed:
Merge branch 'dev' into capstone-meson

XVilka reviewed (approved):
LGTM. But could you please more CI tasks by pushing into the `dist-` branch maybe? So we can check other build jobs too.

ret2libc committed:
Merge branch 'dev' into dist-capstone-meson

ret2libc committed:
Make capstone subproject a C project, not cpp

ret2libc committed:
Merge branch 'dev' into dist-capstone-meson

ret2libc committed:
Fix check_meson_subproject script

ret2libc committed:
Disable unit tests because of bins dependency

ret2libc commented:
**DO NOT MERGE YET, see #510**

ret2libc committed:
Create dist tarball with master meson

Last release 0.56.2 cannot correctly create tarball with subprojects
coming from wrap-git with patches, so let's use an updated meson just
for the `meson dist` job.

ret2libc committed:
Update sdb wrap

ret2libc committed:
Update sdb and pass sdb_includedir option

ret2libc committed:
Merge remote-tracking branch 'origin/dev' into capstone-meson

ret2libc commented:
All green with `dist`, see https://github.com/rizinorg/rizin/pull/510

XVilka reviewed (approved):
LGTM, lets merge this

ret2libc committed:
Merge branch 'dev' into capstone-meson

ret2libc committed:
Mention meson dist in PACKAGER.md
